core/camel-support/src/main/java/org/apache/camel/support/DefaultTimeoutMap.java

```diff
@@ -190,41 +188,18 @@ public void purge() {
             // if we found any expired then we need to sort, onEviction and remove
             if (!expired.isEmpty()) {
                 // sort according to the expired time so we got the first expired first
-                expired.sort(new Comparator<TimeoutMapEntry<K, V>>() {
-                    public int compare(TimeoutMapEntry<K, V> a, TimeoutMapEntry<K, V> b) {
-                        long diff = a.getExpireTime() - b.getExpireTime();
-                        if (diff == 0) {
-                            return 0;
-                        }
-                        return diff > 0 ? 1 : -1;
-                    }
-                });
-
-                List<K> evicts = new ArrayList<>(expired.size());
-                try {
-                    // now fire eviction notification
-                    for (TimeoutMapEntry<K, V> entry : expired) {
-                        boolean evict = false;
-                        try {
-                            evict = onEviction(entry.getKey(), entry.getValue());
-                        } catch (Throwable t) {
-                            log.warn("Exception happened during eviction of entry ID {}, won't evict and will continue trying: {}", 
-                                    entry.getValue(), t);
-                        }
-                        if (evict) {
-                            // okay this entry should be evicted
-                            evicts.add(entry.getKey());
-                        }
-                    }
-                } finally {
-                    // and must remove from list after we have fired the notifications
-                    for (K key : evicts) {
-                        map.remove(key);
-                    }
+                expired.sort(comparing(TimeoutMapEntry::getExpireTime));
+
+                // and must remove from list after we have fired the notifications
+                for (TimeoutMapEntry<K, V> entry : expired) {
+                    map.remove(entry.getKey());
                 }
             }
         } finally {
             lock.unlock();
+            for (TimeoutMapEntry<K, V> entry : expired) {
```

Comment:
This is perhaps the only point of discussion. Notification occurs after eviction and _outside_ of the lock.

Reply:
At first thought I think this is okay, as its only a notification, and you cannot veto an eviction